## Question 2

The table below presents hypothetical bid/ask quotes for four major currencies at a given date/time:

| Pair | Bid | Ask |
| --- | --- | ---- |
| EUR/USD |	0.87060	| 0.87063 |
| GBP/USD |	1.42790	| 1.42796 |
| USD/JPY |	133.330	| 133.333 |
| EUR/GBP |	0.60970	 | 0.60972 |
| EUR/JPY |	116.140 |	116.144 |
| GBP/JPY |	190.480 |	190.476 |

The first currency in a currency pair is called the base currency. The bid and the ask are always quoted in terms of the base currency. For example, if you are buying on EUR/USD trade, then you are effectively receiving EUR in exchange for USD at the ask quote.
  
You were given USD  1  to attempt an arbitrage trade, assume no transaction costs.


### A. Write down a linear programming problem to buy and sell quantities of the base currencies and arbitrage/profit of at most USD  10k .  

First you declare all the variables. D is the number of USD. This is what we will try to maximize. To ensure a profit of at most $10,000, we constrain D to 10,001 (the profit plus the dollar we start with). DE is dollars to euros, DG is dollars to pounds, etc.  
<br>
Then we set up our constraints.  
We start with dollars, we set D = 1 (our starting amt) + the currencies exchanged to dollars - the dollars exchanged to other currencies. We want to end up with dollars.  
<br>
Next we set: the euros we are exchanging = euros we have exchanged for. These should cancel out. We don't end up with any euros.
<br>  
We do the same thing we did for euros to pounds and yen.  
<br>
Finally we set the objective to maximize dollars.

In [10]:
using JuMP, GLPK

m = Model(GLPK.Optimizer)

#Declare Variables:

@variable(m, D <= 10001)

@variable(m, DE >= 0)
@variable(m, DG >= 0)
@variable(m, DJ >= 0)
@variable(m, ED >= 0)
@variable(m, EG >= 0)
@variable(m, EJ >= 0)
@variable(m, GD >= 0)
@variable(m, GE >= 0)
@variable(m, GJ >= 0)
@variable(m, JD >= 0)
@variable(m, JE >= 0)
@variable(m, JG >= 0)

#Declare Constraints

@constraint(m, constraint_1, D == 1 + 0.87060ED + 1.42790GD + (1/133.333)JD - DE - DG - DJ)
@constraint(m, constraint_2, ED + EG + EJ == (1/0.87063)DE + (1/0.60972)GE + (1/116.144)JE)
@constraint(m, constraint_3, GD + GE + GJ == (1/1.42796)DG + 0.60970EG + (1/190.476)JG)
@constraint(m, constraint_4, JD + JE + JG == 133.33DJ + 116.14EJ + 190.48GJ)

#Declare Objective
@objective(m, Max, D)

print(m)

### B. Solve for the buy/sell quantities and the arbitrage/profit amount.

In [11]:
# Solve the optimization problem
JuMP.optimize!(m)

# Print the objective value
println("Objective value: ", JuMP.objective_value(m))

println("D = ", JuMP.value.(D))

println("DE = ", JuMP.value.(DE))
println("DG = ", JuMP.value.(DG))
println("DJ = ", JuMP.value.(DJ))

println("ED = ", JuMP.value.(ED))
println("EG = ", JuMP.value.(EG))
println("EJ = ", JuMP.value.(EJ))

println("GD = ", JuMP.value.(GD))
println("GE = ", JuMP.value.(GE))
println("GJ = ", JuMP.value.(GJ))

println("JD = ", JuMP.value.(JD))
println("JE = ", JuMP.value.(JE))
println("JG = ", JuMP.value.(JG))


Objective value: 10001.0
D = 10001.0
DE = 2.0622362181625046e7
DG = 0.0
DJ = 0.0
ED = 0.0
EG = 0.0
EJ = 2.3686712129865784e7
GD = 0.0
GE = 0.0
GJ = 0.0
JD = 2.7509747467626123e9
JE = 0.0
JG = 0.0


### C. Interpret the results and develop a story of the trade executions and the resulting profit.  

You start with 1 USD.  

You exchange the 1 USD for EUR  

In [12]:
EUR = 1 * (1/0.87063) #(USD -> EUR)
print(EUR)

1.148593547201452

Then you exchange your 1.1486 EUR for JPY  

In [13]:
JPY = EUR * 116.140 #(EUR -> JPY)
print(JPY)

133.39765457197663

Then you exchange your 133.3976 JPY back to USD

In [14]:
USD = JPY * (1/133.333) #(JPY -> USD)
print(USD)

1.000484910502101

You have successfully converted 1 USD into ~1.0005 USD. So your abitrage profit is ~0.0005 USD. If you scale this by 20,622,362.18, you will make a profit of $10,000

In [15]:
20622362.181625046 * (USD-1)

10000.0